### DISCUSSION TITLE
Benutzerverwaltung in Freetz-NG für AVM-Versionen ab 07.2x (07.19-Labor)

### DISCUSSION THREAD

**Discussion by PeterPawn** [Freetz-NG/freetz-ng#261] [Allgemein]
Um die Diskussion zu diesem Thema mal von einem reinen Dialog zu einer mit mehreren möglichen Teilnehmern zu machen, eröffne ich hier dann einfach mal einen Thread dazu.

Es haben sich ja - auch in den letzten AVM-Versionen - ein paar gewichtige Änderungen ergeben, was die Fallstricke bei der Benutzung zusätzlicher (Linux-)Benutzerkonten angeht. Einiges davon wurde mit kürzlich erfolgten Änderungen beim Monitoring von Änderungen der `/var/tmp/passwd` bereits gelöst, aber das erscheint mir persönlich im Moment noch "etwas wackelig" und nachdem jetzt vermutlich das Prinzip erkannt wurde, anhand dessen der `ctlmgr` von AVM Konten übernimmt oder entfernt, wenn er nach Änderungen an der Benutzerverwaltung eine neue `/var/tmp/passwd` erzeugt, kann man das durchaus "stabiler" hinbekommen.

Früher - als die von AVM und von Freetz-NG benutzte C-Library dieselbe Version waren - gab es mit der `libctlmgr` eine Bibliothen, die per Preloading in den `ctlmgr` eingebaut wurde und dessen Umbenennen der neuen `/var/tmp/passwd.tmp` in den richtigen Namen abfing und entsprechend umbog (und letztlich ersetzte, so daß dieses Umbenennen außerhalb der `ctlmgr` fortgesetzt werden mußte).

Das ist bei unterschiedlichen Library-Versionen nicht mehr so einfach, daher wird jetzt die Änderung per INotify-Schnittstelle des Linux-Kernels und mit dem `inotifyd`-Applet der BusyBox überwacht ... nur erfolgt hier die Benachrichtigung erst dann, wenn das Umbenennen schon erfolgt ist und damit die vorhergehende Version der `/var/tmp/passwd` gar nicht mehr verfügbar ist. Daher müssen jetzt die zusätzlichen Konten, die von AVM entfernt wurden, aus einer anderen Quelle wiederhergestellt werden.

Dabei sind mehrere Algorithmen denkbar ... man muß sich also ein paar Gedanken machen, WAS dabei am Ende herauskommen soll und was die verschiedenen Möglichkeiten für Auswirkungen haben - sowohl auf Beschränkungen in den Handlungsoptionen für die Benutzer, als auch auf bestehende Freetz-Installationen, deren Einstellungen mit dem bisherigen Stand eingerichtet wurden und die man nun ggf. anpassen müßte (ob von Hand oder automatisiert, wäre die nächste Frage).

Auf alledem wurde schon einigermaßen ausführlich herumgedacht, nur vermutlich in einer Form, von der die Wenigsten hier etwas mitbekommen haben, nämlich in einem Zwiegespräch in den Kommentaren zu einem einzelnen Commit: https://github.com/Freetz-NG/freetz-ng/commit/4d7bd9c526c8926864337bc8c44e9f56eeb3cc76

Da @cuma/@fda77 das nicht selbst hier eröffnen wollte, mache ich das nun eben ... weitere Kommentare/Beiträge wären hier dann (meines Erachtens zumindest) besser untergebracht, zumal es nur so auch eine Beteiligung/Meinung von tatsächlichen Freetz-NG-**Benutzern** geben kann, denn ich benutze meinerseits ja gar keinen Freetz-Mod selbst (und erst da spielt dann Benutzerverwaltung überhaupt eine Rolle) und @cuma/@fda77 kann das letztlich auch immer nur aus seiner eigenen Perspektive beurteilen.

Niemand von uns beiden weiß aber wirklich, ob es da draußen Leute gibt, die - z.B. für einen FTP-Server, der nicht mit dem `ftpd` von AVM arbeitet - tatsächlich auf die Benutzerverwaltung von Linux angewiesen sind und für die dann Beschränkungen der verwendbaren UIDs für FTP-Benutzerkonten auch Auswirkungen hätten.

Es macht auf der einen Seite wenig Sinn, irgendwelchen Aufwand in ein Feature zu investieren, was letztlich niemand nutzt ... andererseits muß es ja auch nicht zwangsläufig so sein, daß ein Feature, was man selbst "nicht braucht", für andere genauso bedeutungslos und "wenig wünschenswert" ist.

Es braucht also Meinungen und Feedback von denjenigen, für die das letztlich gemacht ist ... und das wären die Freetz-NG-Benutzer. Wer also nicht nur das nutzen will, was einem ohnehin vorgesetzt wird, sondern tatsächlich (bitte sinnvolle) Ideen und Argumente für ein **bestimmtes Vorgehen** hat oder sich auch nur für irgendeinen bereits vorliegenden Vorschlag aussprechen will, der kann dann ja einfach mal (sachlich) aktiv werden.

Ich kann das nicht wirklich in "Wishlist" oder "Feedback" einordnen, daher bleibt's beim "General" als Kategorie. Auch macht es (für mich) keinen Sinn, das bereits im Commit Geschriebene erneut aufzulisten (außer meinem Versuch, den Ausgangszustand zusammenzufassen) und so muß dann jeder halt erst mal im Commit lesen, was die "Alternativen" eigentlich wären. Fühlt sich jemand berufen, auch dafür noch eine Zusammenfassung zu liefern ... unter diesem Beitrag ist ein Feld, wo "Write a reply ..." stehen sollte.

**Comment by fda77** [comment]
Dass die AVM-Benutzer heraus- bzw filtert wurden war mit libctlmg auch schon so. Da fiel zuletzt auf als die appuser dazukamen. Ansonsten ist das kein Problem.

In einem uralten Backup FOS 04.89 hab ich diese passwd
```
root:x:0:0:root:/mod/root:/bin/sh
ftpuser:x:100:1:ftpuser:/var/media/ftp:/bin/false
ftp:x:101:1:FTP account:/home/ftp:/bin/false
davfs2:x:102:1000:DavFS2:/home/davfs2:/bin/false
nobody:x:103:1002:nobody:/home/nobody:/bin/false
ntp:x:104:1003:ntp:/home/ntp:/bin/false
openvpn:x:1000:1001:OpenVPN account:/home/openvpn:/bin/false
```

Anderes gerät mit FOS 04.34
```
root:x:0:0:root:/mod/root:/bin/sh
ftp:x:100:1:FTP account:/home/ftp:/bin/false
ftpuser:x:101:1:ftpuser:/var/media/ftp:/bin/false
```

Wie diese IDs zustande kamen weiss ich nicht.
In welchem Bereich man am besten zukünftig neue User anlegen sollte auch nicht

Wenn du gerne neue Kategorien möchtest kann ich dir auch nur Moderatoren-Berechtigungen geben, da du ja keine commit Rechte haben willst

**Reply by PeterPawn** [reply]
Ich weiß zwar, daß ich meinen guten Vorsätzen, mich nicht mehr einzumischen, immer wieder untreu werde ... aber so, wie es jetzt ist, ist es für mich auch in Ordnung. Mit irgendwelchen zusätzlichen Rechten würde ich mir nur (in meiner Welt) auch mehr Verantwortung aufladen, auch dann "zu reagieren", wenn mich irgendetwas absolut nicht tangiert ... und ich habe ja immer noch eigene Pläne, die nun so gar nicht zu Freetz-NG und/oder Freetz passen - auch mein eigener Fork ist/war ja nur Mittel zum Zweck.

Lassen wir es bei der "voice from the off" ...

---

> Dass die AVM-Benutzer heraus- bzw filtert wurden war mit libctlmg auch schon so.#

Das mag ja sein ... trotzdem hatte man dann noch die Chance (oder hätte sie gehabt, wenn man das so implementiert hätte), die von AVM gelöschten Accounts direkt aus der weiterhin noch existierenden `/var/tmp/passwd` dazu zu mischen, solange das Umbenennen NICHT ausgeführt wurde. 

Damit würden dann auch Konten überleben, die zuvor in der `passwd` standen und trotzdem noch nicht mit einem `modusers save` in die Freetz-Einstellungen übernommen wurden. 

Wenn das bisher nicht so gehandhabt wurde ... auch gut - es geht mir ja auch darum, was man da generell ändern kann/sollte (u.U. sogar noch für Boxen, die nur 07.0x erhalten haben ... auch deren Besitzer werden sicherlich früher oder später mal auf neuere Modelle migrieren müssen), was man dabei für Optionen hat, was man berücksichtigen muß und welche Auswirkungen welche Entscheidung dann hat. 

Letztlich als - hier "im Kleinen" probiert - mal so etwas wie das gemeinsame Entwickeln eines "Plans", wohin die Reise gehen soll. Wer das dann am Ende implementiert und ob das überhaupt jemand macht, wäre erst der nächste Schritt - auch wer das tatsächlich per PR als Änderung vorschlagen wollte, sollte sich ja erst mal einen Überblick verschaffen, was "die Leute" (als Zielgruppe) eigentlich brauchen würden und was er/sie dann letztlich implementieren müßte.

**Comment by fda77** [comment]
Es war ja bisher so dass man /etc/passwd und die anderen ändern konnte wie man wollte, und danach ein "modusers save" ausführen musste. Danach noch ein "modsave flash" oder ganz einfach nur ein "modsave". Es ist halt schwer allen zu vermitteln dass das nun anders laufen soll. Vor allem wenn es nicht bei allen FOS so ist...
Beispiel:
https://github.com/Freetz-NG/freetz-ng/issues/236#issuecomment-804280949
https://github.com/Freetz-NG/freetz-ng/issues/236#issuecomment-804315890
https://freetz-ng.github.io/freetz-ng/wiki/10_Beginner/newbie_other.html#ftp-user-erzeugen

Aktuell ist es so dass man modaddusers bei neuen FOS verwenden muss (und bei allen kann). Vielleicht würde man das busybox Applet doch besser umbenennen und direkt durch das Script ersetzen, dann wäre es absolut gleich beim bedienen

Aber viel wichtiger ist die Sache mit den IDs!

**Comment by fda77** [comment]
Eigentlich wäre ein Issue besser gewesen.
Ich hab mal alle letzten Änderungen wegen passwd rückgängig gemacht und LAST_SYSTEM_ID von 999 auf 899 gesetzt. Neue Benutzer blieben so bei Reboot und Änderungen im AVM-Webif erhalten.  Man braucht dann kein Inotify-Prozess und die Bedienung bleibt wie sie war. Bestehende user muss man gar nicht ändern, da sie eh gelöscht wurden :)

**Reply by fda77** [reply]
Keine Einwände seh ich als Zustimmung an
